src/host.rs

```diff
@@ -97,29 +255,81 @@ impl Host {
 
     pub(crate) fn recv_from(&mut self, peer: SocketAddr) -> (Option<Envelope>, Rc<Notify>) {
         let now = Instant::now();
-        let notify = self.notify.clone();
 
         let deque = self.inbox.entry(peer).or_default();
+        let notify = self.notify.clone();
 
         match deque.front() {
             Some(Envelope {
                 instructions: DeliveryInstructions::DeliverAt(time),
                 ..
             }) if *time <= now => {
-                self.version += 1;
-                (deque.pop_front(), notify)
+                let ret = (deque.pop_front(), notify);
+                self.bump_version();
+                ret
             }
             _ => (None, notify),
         }
     }
 
+    pub(crate) fn recv_on(&mut self, pair: SocketPair) -> Option<Segment> {
+        let now = Instant::now();
+
+        let deque = self
+            .connections
+            .get_mut(&pair)
+            .map(|i| &mut i.deque)
+            .expect("no connection");
+
+        match deque.front() {
+            Some(StreamEnvelope {
+                instructions: DeliveryInstructions::DeliverAt(time),
```

Comment:
Unlike the other comment (for accepting connections), I think this code is correct. Just flagging it in case I'm wrong.

Reply:
Yup, we've got ordering guarantees within the connection.